### ISSUE TITLE
An Empirical Study of the Generalization Ability of Lidar 3D Object Detectors to Unseen Domains

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#59]
<img width="1185" alt="image" src="https://github.com/user-attachments/assets/8eec6658-8f5e-431a-88ad-8cd31b4b0833">

<img width="500" alt="image" src="https://github.com/user-attachments/assets/e2f20d3f-9a6e-4594-bd34-fb82365a1310">

### Paper Link
https://openaccess.thecvf.com/content/CVPR2024/html/Eskandar_An_Empirical_Study_of_the_Generalization_Ability_of_Lidar_3D_CVPR_2024_paper.html

## Motivation

<img width="500" alt="image" src="https://github.com/user-attachments/assets/1061a71f-f887-4315-9ee9-42d8d6c5a72c">

- 3D detectors perform **poorly** on domains they were **not trained** on, i.e. different locations, sensors, weather, etc., limiting their reliability in safety-critical applications. 
- There exist methods to adapt 3D-ODs to these domains; however, these methods treat 3D-ODs as a **black box**, neglecting underlying architectural decisions and source-domain training strategies. 
- Instead, we dive deep into the details of 3D-ODs, focusing our efforts on **fundamental factors** that influence robustness prior to domain adaptation.

We systematically investigate four design choices (and the interplay between them) often overlooked in 3D-OD robustness and domain adaptation: **architecture, voxel encoding, data augmentations, and anchor strategies**.  We assess their impact on the robustness of nine state-of-the-art 3DODs across six benchmarks encompassing three types of domain gaps - **sensor type, weather, and location**. 

## Experimental Setup
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/b1b911f5-aabd-4259-9269-8bf5e3b483d0">
<img width="500" alt="image" src="https://github.com/user-attachments/assets/4d591054-ad3e-4a35-aed9-fe7168c4b55c">

## Main Results and Key Findings

<img width="1590" alt="image" src="https://github.com/user-attachments/assets/45aa2184-fe63-4d95-aec3-61b38e14b9e9">

1. **Transformers vs Non-transformers.** VOTRs outperform 3D CNNs in the mean AP across all domain gaps _when coupled with point features_ (VOTR-TSD). 3D CNNs can be more robust than VOTRs in _voxel-only_ models (VoxelRCNN versus VOTRVoxelRCNN). **Reason**: The addition of point features enhances the performance of VOTR, as it adds a much-needed spatial local context to the transformer’s large receptive field of view.
2. **Anchorless** detectors are robust in the _weather_ domain gap. This is more evident on the _Pedestrian_ class, where CenterPoint and PVRCNN-Centerhead outperform other models by 4-5 AP.
3. **Adding point features in the backbone increases robustness, especially for transformers.** VOTRTSD outperforms VOTR-VoxelRCNN (Tab. 3), PVRCNNCenterhead outperforms Centerpoint but PVRCNN and VoxelRCNN have similar performance. The key distinction between these pairs is point features in the backbone.

## Location Domain Gap

Discrepancies in **object sizes across geographical locations** are the main cause of bad generalization.

### Anchor Size 

The anchor size in 3D-OD is a carefully tuned hyperparameter and is usually close to the mean object size (MOS) of the training dataset. We seek to answer two questions: 

1. Is there a connection between anchor size and performance in OOD scenarios? 
2. And, when considering training and test phases, which holds greater significance: the anchor size during training or testing?

<img width="500" alt="image" src="https://github.com/user-attachments/assets/d0b3d798-0538-43f3-a819-1ffafad7d644">

1. There is **no correlation** between the anchor size at training time and the OOD performance. All three models have low AP when they use the same anchor size they were trained on.
2. Changing anchor size **at test time** leads to strong variation in the performance.
3. The best performing anchors are found to be **always smaller than the training anchor**. 

> object sizes of KITTI are smaller than Waymo?

<img width="500" alt="image" src="https://github.com/user-attachments/assets/dd0cd848-67db-4415-affe-fa86c780629e">

**Tuning the Anchor Size on Target Datasets with Larger Objects.** In Tab. 11, we report the effect of changing the anchor size of SECOND on N→W, where the target objects in this setting are larger than the source objects. We find that the best-performing anchors are still smaller than the training anchors.

### Anchor vs. Label Strategy

<img width="500" alt="image" src="https://github.com/user-attachments/assets/493fd487-91f2-4c62-aeec-f47cdc4d9a83">

1. **Anchorless** detectors exhibit **only a marginal improvement** in generalization compared to anchor-based detectors when dealing with **objects of various sizes**. Their performance is notably suboptimal.
2. ROS enhances robustness on KITTI but **decreases performance on NuScenes and the source domain.** 
3. The utilization of multiple anchor sizes adds positive gains on Waymo and W→N.

**There is no universally generalizable technique at training time across all locations.** From a continual learning perspective, we argue it is more beneficial to **fix the original training labels and change the anchor at test time** using a small set of labeled target data.

### Effects on different Architectures

We evaluate the effect of **tuning the anchor size at test time** on seven state-of-the-art anchor-based models.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/f68e04e3-7ff9-4958-b42a-079ad86ec4f8">

1. The performance of all models increases on all 3 benchmarks. 
2. The effect is more pronounced when the discrepancy between object sizes in the source and target datasets is great. For instance, the effect is stronger on W→K than on W→N. 
3. This technique more positively influences **voxel-only** models than point-only or hybrid models.

> Can we design a strategy to adaptively tune the anchor size at test-time for adatptation?

## Sensor Domain Gap

We follow [43] and study the benchmarks which predominantly exhibit a **discrepancy on the sensor level**, namely K64→K32, K64→K16, K16→K64 and W→N (since the average object sizes between Waymo and NuScenes are close). We make the distinction between two cases: (1) same Vertical-FOV and **different number of beams**, and (2) **different V-FOV**. The second case is represented by W→N, resulting in a different sampling pattern and is considerably harder to solve.

### Low-to-High vs. High-to-Low

<img width="1834" alt="image" src="https://github.com/user-attachments/assets/ec690151-a2ed-418d-96fe-2ff52a561c89">

High-to-low domain gap is much harder to tackle than the low-to-high gap.

> Most of DA works focus on high-to-low.

### Voxel Encoding

<img width="500" alt="image" src="https://github.com/user-attachments/assets/cdaedc71-35c8-4aca-85be-5afc971dada5">

- A common view in the field [43] is that **a smaller number of points per foreground object at test time is behind the performance drop**.
- A smaller number of beams results in **more empty voxels vertically**, which will not be used for detection.
- On the other hand, a **larger** height would **reduce the difference in the number of empty voxels** between source and target domains.

<img width="1778" alt="image" src="https://github.com/user-attachments/assets/2054107b-af26-44b0-bd7a-da064184f821">

There is a **positive** correlation between **voxel height and robustness on low resolution sensors,** as the 3D mAP increases in most experiments on the target domain (K32 and K16) while remaining relatively stable on the source domain (K64).

### Data Augmentations 

- GT-sampling: augments the number of foregrounds by drawing samples from a database of source domain objects.
- Shape augmentation (SA): removes certain parts of the object and/or downsamples its points.
- Line downsampling (LD): inspired by [43], consists in downsampling Lidar beams of the source scans in the vertical direction.
- A variant of GT-sampling, which we term Mixed GT-Sampling, consists in adding samples extracted from another dataset with a different V-FOV compared to the source dataset. In this case, we add samples from KITTI for W→N and samples from NuScenes for the KITTI benchmarks.

<img width="1540" alt="image" src="https://github.com/user-attachments/assets/11c63121-180f-4b79-8db0-761dc6af7543">

1. GT-Sampling has **negative** effects on some classes, as the detection accuracy deteriorates on pedestrians in K64→K32-16 and cars in W→N. We hypothesize this is due to **overfitting on source** domain shapes.
2. Mixed GT-Sampling can **restore** the performance drop of GT-Sampling and adds small gains, reducing the overfitting of the original method.
3. The impact of point sparsity augmentations on robustness is **the most significant**.

## Weather Domain Gap

The weather domain gap in Lidar is **particularly hard** to address: it introduces artifacts such as **missing points or clutters resulting from light scattering**. Sometimes a significant part of the pointcloud can be missing.

### Transferability versus Discriminability

We notice that a simple in-domain evaluation of common 3D detectors on bad weather data is missing. In Figure 7, we assess different models using the Kirkland validation split. Each model is trained two times, once on Waymo data (W→Kr) and once on the Kirkland training split, which is regarded as the oracle performance or upper bound.

<img width="1550" alt="image" src="https://github.com/user-attachments/assets/84341281-ee72-4b3d-a38b-ef0ce7b25a3b">

**Findings**: Models trained on clean weather data exhibit comparable or even superior generalization to poor weather conditions compared to those trained on bad weather data. This is especially true for the class Pedestrian, which exhibits the most deterioration in adverse weather.

This reveals that detection in bad weather conditions is **not merely a transferability problem**; rather, it is **a discriminability problem** where simple supervised learning is hard. We believe more research should focus on building better models for this domain or on integrating other modalities (camera, radar) to improve detection in adverse conditions.

### Data Augmentations and Voxel Size

<img width="500" alt="image" src="https://github.com/user-attachments/assets/a4d11eab-bc35-492a-98bb-120b2c22ccc1">

1. The weather domain gap is not significantly affected by augmentations.
2. PVRCNN with a smaller voxel size trained only on 20% of Waymo dataset is able to outperform SPG [44] on the Vehicle class, as SPG uses a larger voxel size. This highlights again the importance of choosing the right voxel size as it can increase or decrease the model’s robustness.
3. Training on source and target does not improve the performance. 

### Anchor Size
1. Changing anchor size at test-time can notably enhance model performance **across various locations**.
2. **Training with larger anchors is recommended** for practical applications, as the optimal anchor is always found to be smaller than the training anchor. This can enable semi-supervised, federated, and test-time DA without model retraining, allowing a seamless transition between locations by only changing the anchor.

### Architecture and Voxel Size
1. Point-voxel representations in the backbone show better robustness across most domain pairings.
2. Voxel transformer backbones are more robust than 3D CNNs when coupled with **point features**.
3. Anchorless detectors are the most robust in **adverse weather**.
4. While shorter voxel heights generally improve performance on the source domain, **increasing voxel height improves the robustness** when training on high-resolution sensors and inferring on low-resolution sensors (**high-to-low**).

### Data and Augmentations
1. Ground-truth sampling has **negative** effects on the robustness on lower-resolution sensors, likely due to overfitting on the source domain shapes. Adding groundtruth samples from another dataset with a different V-FOV reduces the overfitting.
2. **Point sparsity augmentations**, like shape and line downsampling augmentations, prove effective on **high-to-low** resolution domain gaps.
3. Going from high-to-low resolution is more challenging than the reverse due to point sparsity at test-time, with models performing better on higher-resolution test data.
4. Surprisingly, training on clean weather samples leads to more robustness on bad weather than direct training on bad weather or mixed samples. This highlights the importance of choosing the right training data but also points out that the problem with bad weather data is about discriminability rather than transferability.

## Notes

In my TTA-3OD research, 
- For  **cross-location** domain gap, smartly adusting anchor size at test-time may be a solution.
- For **high-to-low** resolution domain gaps, we still needs point sparsity augmentations. 
- For **weather** domain gaps, need to train on high-quality datasets.
